docs/design-documents/promtail-scraping-objectstore.md

```diff
@@ -0,0 +1,103 @@
+
+Cortex already has the [implementation](https://github.com/cortexproject/cortex/blob/master/pkg/chunk/aws/s3_storage_client.go) for Objecstore including S3 and GCS. We can try to re-use most of it. Still, we have to develop code for `file watching` and updating the `positions.yaml` file.  Following are the few challenges which we should try to address.
+
+### There is no appending of lines to file in object store
+
+In an object store there is no appending of lines to a file in bucket. Any file modified will be newly 
+uploaded to the bucket with the same name. So, we would have to read the entire file again from the begining since we will not be aware if only new lines were appended to the modified file or the file is completely having new content
+
+### Where should we keep positions.yaml file ?
+
+Since, the files we scrape is remotely placed in a objecstore, should we keep the `positions.yaml` also in the same place? Advantage of keeping the `positions.yaml` within the objecstore will help us to recover from the state where the scraping was left when promtail is running in docker container and the container crashes. However, `user` reading the files from the Object store would also need write access to the Objecstore. Getting write access may not be possible in some cases where people doesn't want to risk modifying the files by giving write access.
+
+### How should we support reading of compressed files ?
+
+If the files in the Objectstore is stored in a compressed format, promtail should have support for reading the compressed data and we should make sure we support most of the compression algorithms like we already do for chunks(Snappy, LZ4, Gzip) 
+
+### How we should identify and watch newly added files ?
+
+To identify and to start tailing the newly added files, we should list the files in the Object store at some desired intervals. How frequently we should this? Once we add the file to `watch`, to identify any changes to the file we should frequently check the modified `datetime` of the file. The list of files might be large and each watch will be running in a `go routine` and each `go routine` making frequent `http` requests to the Obejctstore. We should make sure this doesn't affect promtail's performance.
+
+Instead of frequently doing `ListObjects`, we can consider to use event notification in S3 with `SQS`, through which we can get to know the newly added files and then we can fetch only them.
+
```

Comment:
@cyriltovena WDYT?

Reply:
Let me put few words please.
SQS doesn't support FIFO for event notifications from S3. That means we will have "out of order" errors in Loki.

We're using S3, SQS, fluentd and loki now and sometimes get this error. AWS has no plan to support FIFO for S3 notifications, we asked them.

Reply:
> SQS doesn't support FIFO for event notifications from S3. That means we will have "out of order" errors in Loki.

Nice point. 

> AWS has no plan to support FIFO for S3 notifications, we asked them

This is very good information. It's also mentioned here https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html

Thanks @Punkoivan